pkg/extensions/tbtc/tbtc.go

```diff
@@ -0,0 +1,286 @@
+		keepClosedChan, keepClosedUnsubscribe, err := keepClosedFn(deposit)
+		if err != nil {
+			logger.Errorf(
+				"could not setup keep closed handler for [%v] "+
+					"monitoring for deposit [%v]: [%v]",
+				monitoringName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer keepClosedUnsubscribe()
+
+		timeoutChan := time.After(timeout)
+
+		actionAttempt := 1
+
+	monitoring:
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Infof(
+					"context is done for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-stopEventChan:
+				logger.Infof(
+					"stop event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-keepClosedChan:
+				logger.Infof(
+					"keep closed event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-timeoutChan:
+				err := actFn(deposit)
+				if err != nil {
+					if actionAttempt == maxActionAttempts {
+						logger.Errorf(
+							"could not perform action "+
+								"for [%v] monitoring for deposit [%v]: [%v]; "+
+								"last attempt failed",
+							monitoringName,
+							deposit,
+							err,
+						)
+						break monitoring
+					}
+
+					backoff := actBackoffFn(actionAttempt)
+
+					logger.Errorf(
+						"could not perform action "+
+							"for [%v] monitoring for deposit [%v]: [%v]; "+
+							"retrying after: [%v]",
+						monitoringName,
+						deposit,
+						err,
+						backoff,
+					)
+
+					timeoutChan = time.After(backoff)
+					actionAttempt++
+				} else {
+					logger.Infof(
+						"action for [%v] monitoring for "+
+							"deposit [%v] has been performed successfully",
+						monitoringName,
+						deposit,
+					)
+					break monitoring
+				}
+			}
+		}
+
+		logger.Infof(
+			"[%v] monitoring for deposit [%v] "+
+				"has been completed",
+			monitoringName,
+			deposit,
+		)
+	}
+
+	return monitoringStartFn(
+		func(deposit string) {
+			go handleStartEvent(deposit)
+		},
+	)
+}
+
+func (t *tbtc) watchKeepClosed(
+	deposit string,
+) (chan struct{}, func(), error) {
+	signalChan := make(chan struct{})
+
+	keepAddress, err := t.chain.KeepAddress(deposit)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	keepClosedSubscription, err := t.chain.OnKeepClosed(
+		common.HexToAddress(keepAddress),
+		func(_ *chain.KeepClosedEvent) {
+			signalChan <- struct{}{}
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	keepTerminatedSubscription, err := t.chain.OnKeepTerminated(
+		common.HexToAddress(keepAddress),
+		func(_ *chain.KeepTerminatedEvent) {
+			signalChan <- struct{}{}
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	unsubscribe := func() {
+		keepClosedSubscription.Unsubscribe()
+		keepTerminatedSubscription.Unsubscribe()
+	}
+
+	return signalChan, unsubscribe, nil
+}
+
+func exponentialBackoff(iteration int) time.Duration {
```

Comment:
Can we add a comment explaining how the first backoffs will look like? My understanding is:
```
// 2000-2100ms,  4000-4100ms, 8000-8100ms,...
```